docs: add peer discovery and connection documentation (#169)
tweag/cardano-hoarding-node: docs/peer-discovery-and-connection.md

### Core Modules

| Module                                    | Responsibility                                                          |
| ----------------------------------------- | ----------------------------------------------------------------------- |
| `Hoard.Bootstrap`                         | Initial peer bootstrapping from configuration                           |
| `Hoard.Collector`                         | Per-peer connection lifecycle management                                |
| `Hoard.Effects.NodeToNode`                | Network connection and protocol implementations (including PeerSharing) |
| `Hoard.Listeners.DiscoveredNodesListener` | Spawn collectors for newly discovered peers                             |
| `Hoard.Effects.PeerRepo`                  | Peer database operations                                                |

## Bootstrap Process

**File:** `src/Hoard/Bootstrap.hs`

On first startup (when no peers exist in the database), the node bootstraps from the configuration:

```haskell
bootstrapPeers :: (PeerRepo :> es, Clock :> es, IOE :> es, Reader Config :> es)
               => Eff es (Set Peer)
```

> ### Bootstrap Steps

Reviewer: I don't know whether this heading is rendered correctly since it's in a different color to the rest of the headings here 🤷

1. **Extract relay addresses** from `bigLedgerPools` in the peer snapshot configuration
2. **Resolve domains/IPs** to concrete `PeerAddress` values:
   - For domain-based relays: resolve DNS to IP addresses
   - For IP-based relays: parse IP directly or resolve as hostname if parsing fails
3. **Persist to database** via `upsertPeers` with the first address as the bootstrap source
4. **Return peer set** for initial collector spawning

**Configuration Source:** The bootstrap configuration is provided through the `peerSnapshot` field in the application config, which contains `bigLedgerPools` - a list of well-known stake pool relays.

## Connection Lifecycle

> ### Collector Spawning

Reviewer: I don't know whether these headings are rendered correctly since they're in a different color to the rest of the headings here 🤷

```

**Phases:**

1. **Acquire** (`src/Hoard/Collector.hs:102-115`):
   
   - Check if peer is already connected (tracked in `HoardState.connectedPeers`)
   - Check if peer is eligible for connection based on cooldown period
   - If eligible and not connected: add peer ID to `connectedPeers` state
   - If ineligible or already connected: skip

2. **Use** (`src/Hoard/Collector.hs:116-117`):
   
   - Execute `runCollector` to spawn protocol mini-protocols
   - This runs until connection fails or is terminated

3. **Release** (`src/Hoard/Collector.hs:118-134`):
   
   - On exception: check if it's a graceful shutdown
   - If **not** graceful: call `updatePeerFailure` to record failure timestamp
   - Remove peer ID from `connectedPeers` state
   - Publish `CollectorStopped` event

### Peer Eligibility and Retry Logic

**File:** `src/Hoard/Collector.hs`

```haskell
isPeerEligible :: UTCTime -> Peer -> Bool
isPeerEligible currentTime peer =
    case peer.lastFailureTime of
        Nothing -> True  -- Never failed, always eligible
        Just failureTime ->
            let timeSinceFailure = diffUTCTime currentTime failureTime
            in timeSinceFailure >= peerFailureCooldown

> 

Reviewer: ```suggestion
            in timeSinceFailure >= peerFailureCooldown
```

The suggestion view here breaks a bit, but I'm suggesting trimming the two empty lines 123 and 124.

└────────────────────┬────────────────────────────────────────┘
                     │
                     │ For each new peer
                     │
                     ▼
┌─────────────────────────────────────────────────────────────┐
│ bracketCollector (recursive spawning)                       │
│                                                              │
│ 1. Check eligibility (not in cooldown)                      │
│ 2. Add to connectedPeers state                              │
│ 3. Run collector → spawn all protocols                      │
│ 4. PeerSharing in new collector will discover MORE peers    │
│ 5. Recursive growth of peer network                         │
└─────────────────────────────────────────────────────────────┘
```

### Peer Persistence

**File:** `src/Hoard/Effects/PeerRepo.hs:97-117`

```haskell
upsertPeers :: Set PeerAddress -> PeerAddress -> UTCTime -> Eff es (Set Peer)
```

**Logic:**

1. For each peer address in the set:
   - Check if a peer with this address already exists in the database
   - If exists: skip (no update needed)
   - If new: insert new peer record with:
     - `discoveredFrom`: the source peer address

>      - `discoveredAt`: discovery timestamp

Reviewer: We don't have these columns in our `peers` table at all though. We have `discoveredVia` instead of `discoveredAt`, but we have no analog for `discoveredFrom`.

     - `lastFailureTime`: NULL (no failures yet)
2. Return the set of **newly created** peers only

> **Result:** Only new peers are returned, preventing duplicate collector spawning for already-known peers.

Reviewer: I think both new and updated peers are returned. According to [Rel8's docs](https://hackage-content.haskell.org/package/rel8-1.7.0.0/docs/Rel8.html#t:Returning), `Returning` projects out the affected rows.

### Cooldown Period

**Duration:** 5 minutes (configurable)

**Enforcement:** Before attempting connection in `bracketCollector`, `isPeerEligible` checks if sufficient time has passed since `lastFailureTime`.

**Benefits:**

- Prevents rapid reconnection attempts to failing peers
- Reduces network churn and resource usage
- Allows transient issues to resolve before retry
- Respects peers that may be experiencing problems

### Connection State Management

**State Location:** `HoardState.connectedPeers :: Set PeerId`

**Purpose:** Track which peers currently have active connections to prevent duplicate connection attempts

**Lifecycle:**

- **Add**: When `bracketCollector` successfully starts (`src/Hoard/Collector.hs:108-112`)
- **Remove**: In `bracketCollector` exit handler, regardless of success/failure (`src/Hoard/Collector.hs:135-142`)

**Thread Safety:** Uses `Effectful.State.Static.Shared.State` which provides thread-safe state management across concurrent collectors.

## ## Database Schema

### Peers Table

> **File:** `src/Hoard/DB/Peers/Schema.hs`

Reviewer: ```suggestion
**File:** `src/Hoard/DB/Schemas/Peers.hs`
```

```haskell
data Peers f = Peers

>     { id :: Column f PeerId

Reviewer: ```suggestion
    { id :: Column f (ID Peer)
```

    , host :: Column f NodeIP

>     , port :: Column f Word16

Reviewer: ```suggestion
    , port :: Column f Int32
```

    , discoveredFrom :: Column f NodeIP
    , discoveredFromPort :: Column f Word16
    , discoveredAt :: Column f UTCTime
    , lastFailureTime :: Column f (Maybe UTCTime)  -- NULL if never failed
    }
```

**Key Fields:**

- `id`: Unique peer identifier (UUID)
- `host`, `port`: Peer's network address
- `discoveredFrom`, `discoveredFromPort`: Source peer that shared this address
- `discoveredAt`: When this peer was first discovered
- `lastFailureTime`: Most recent connection failure timestamp (NULL = never failed or last attempt succeeded)

### Important Operations

> **Upsert Peers:** Insert new peers, skip existing ones by address

Reviewer: ```suggestion
**Upsert Peers:** Insert new peers, update `lastSeen` on existing ones by address
```

**Update Failure:** Set `lastFailureTime` to current timestamp
**Get All Peers:** Retrieve all known peers for collector spawning

## ## Performance Considerations

### Connection Limits

**Current Implementation:** No hard limit on concurrent connections

> **State Tracking:** `HoardState.connectedPeers` tracks active peer IDs but doesn't enforce a maximum

Reviewer: This does not actually explain any particular performance implications with having too many concurrent connections. Should it?

### Cooldown Impact

**Benefit:** Prevents retry storms from failing peers
**Trade-off:** May slow peer network growth if many initial peers are unreachable
**Configuration:** Currently hardcoded to 5 minutes; will be made configurable

### PeerSharing Request Rate

**Current:** Requests 100 peers every 1 hour per connection
**Rationale:**

- Avoids overwhelming peers with requests
- Provides continuous peer discovery

> - 100 peers per request is the protocol maximum

Reviewer: I thought `10` was the maximum? I might be wrong on that though.

Author: We should make this configurable as well, by the way.

**Growth Model:** With N connected peers, we discover up to 100 * N new addresses per hour (deduplicated by database)

### Database Deduplication

**Efficiency:** `upsertPeers` uses address-based uniqueness to prevent duplicate peer records

**Benefit:** Multiple peers may share the same address; database ensures we only store and connect once per unique address

## Summary

The peer discovery and connection system provides:

> - **Automated Bootstrap**: Start from hardcoded relays, no manual peer management required

Author: Reworded :+1:

> - **Recursive Discovery**: Exponential peer network growth through PeerSharing protocol requesting 100 peers/hour from each connected peer

Reviewer: Referring to the "I don't think `bracketCollector` is recursive" comment further up, _this_ is the recursive description I was looking for!
